modules/argonaut/src/main/scala/xenomorph/argonaut/FromJson.scala

```diff
@@ -79,6 +79,9 @@ object FromJson {
 
         case IsoSchema(base, iso) =>
           base.map(iso.get)
+
+        case PrismSchema(base, prism) =>
```

Comment:
> I had previously hesitated to add this as a schema constructor, because it's not possible to give a good error message

That's a valid point. In my formulation library you have to return an `Attempt` which can be converted from `Option`, `Try` and `Either` http://vectos.net/formulation/docs/richer-data-types.html#attempt

This works pretty well for example on a enumeration:


```scala
trait Enum[A] {
  val allValues: Set[A]
  def asString(value: A): String
}

sealed abstract class Color(val repr: String)

case object Black extends Color("black")
case object White extends Color("white")
case object Orange extends Color("orange")

val all: Set[Color] = Set(Black, White, Orange)

implicit val enum: Enum[Color] = new Enum[Color] {
    val allValues = all
    def asString(value: Color): String = value.repr
}

def enum[A](implicit E: Enum[A]): Avro[A] =
    string.pmap(str => Attempt.fromOption(E.allValues.find(x => E.asString(x) == str), s"$str not found"))(E.asString)
```
>  I'm not sure what it would mean for something like JSON-Schema generation; essentially, this is saying that there's some part of the data which is not adequately described by the schema.

Recently I worked on a JSON library which adds more refinement of primitives types. Thats could be a direction to look into:

```scala

implicit val refinedClass: Jason[RefinedClass] = object4("RefinedClass")(RefinedClass.apply)(
    "name" -> member(string.matchesRegex, _.name),
    "uri" -> member(string.url, _.uri),
    "age" -> member(int.positive, _.age),
    "hue" -> member(int.intervalOpen, _.hue)
  )

case class RefinedClass(
  name: String Refined MatchesRegex[W.`"[A-Za-z+]"`.T],
  uri: String Refined Url,
  age: Int Refined Positive,
  hue: Int Refined Interval.Open[W.`1`.T, W.`10`.T]
)
```

It actually extracts the information from the refinement types to emit into the `SchemaF` types:
https://github.com/vectos/jason/blob/master/refined/src/main/scala/jason/refined/package.scala